src/Map.elm

```diff
@@ -138,92 +143,134 @@ type Action
 
 
-landscapes : Array ( LandscapeName, LandscapeResources )
-landscapes =
-    Array.fromList
-        [ ( "Field", [ ( Pluck, flowers ) ] )
-        , ( "Forest", [ ( Chop, woodResources ), ( Pluck, mushrooms ) ] )
-        , ( "Hill", [ ( Mine, rockResources ) ] )
-        ]
+type LandscapeName
+    = LandscapeName String
 
 
-findSafeInDict : Int -> ResourceDict a -> a
-findSafeInDict id dict =
-    case Array.get id dict of
-        Just res ->
-            res
+type alias Landscape =
+    ( LandscapeName, LandscapeResources )
 
-        Nothing ->
-            Debug.todo ("Didn't find " ++ String.fromInt id ++ " in " ++ Debug.toString dict)
 
+type Landscapes
+    = Landscapes (Array Landscape)
 
-findSafeInList : Int -> List a -> a
-findSafeInList id list =
-    case Array.fromList list |> Array.get id of
-        Just el ->
-            el
 
-        Nothing ->
-            Debug.todo ("Didn't find " ++ String.fromInt id ++ " in " ++ Debug.toString list)
+landscapes : Landscapes
+landscapes =
+    Landscapes
+        (Array.fromList
+            [ ( LandscapeName "Field"
+              , LandscapeResources
+                    [ LandscapeResource ( Pluck, flowers ) ]
+              )
+            , ( LandscapeName "Forest"
+              , LandscapeResources
+                    [ LandscapeResource ( Chop, woodResources )
+                    , LandscapeResource ( Pluck, mushrooms )
+                    ]
+              )
+            , ( LandscapeName "Hill"
+              , LandscapeResources
+                    [ LandscapeResource ( Mine, rockResources ) ]
+              )
+            ]
+        )
 
 
-generateLocationData : Int -> LandscapeId -> Maybe LocationData
-generateLocationData seed (LandscapeId landscapeId) =
+generateLocationData : LandscapeId -> Generator (Maybe LocationData)
+generateLocationData (LandscapeId landscapeId) =
     let
-        ( _, landscapeValue ) =
-            findSafeInDict landscapeId landscapes
-
-        ( randomLandscapeResourceTypeId, resourceTypeSeed ) =
-            step (Random.int 0 (List.length landscapeValue - 1)) (initialSeed seed)
+        (Landscapes l) =
+            landscapes
 
-        ( landscapeAction, landscapeResources ) =
-            findSafeInList randomLandscapeResourceTypeId landscapeValue
-
-        ( resourceId, resourceSeed ) =
-            step (Random.int 0 (Array.length landscapeResources - 1)) resourceTypeSeed
+        ( _, landscapeResources ) =
+            findSafeInDict landscapeId l
+    in
+    genRandomLandscapeResource landscapeResources
+        |> Random.andThen
+            (\(LandscapeResource ( resourceAction, resources )) ->
+                genRandomResource resources
+                    |> Random.andThen
+                        (\(Resource ( ResourceName resourceName, Rarity rarity, maxAmount )) ->
+                            genAmountChance maxAmount
+                                |> Random.andThen
```

Comment:
If you use `Random.map` instead of `Random.andThen` here you won't have to wrap the results of the expression in a `Random.constant`

Reply:
Good catch! Thank you!